### DISCUSSION TITLE
Load log4j.xml file

### DISCUSSION THREAD

**Discussion by riwarek** [apache/logging-log4j2#2186] [Q&A] [answered]
I am during process of update my logging system from verions 2.2 to 3.0.0-beta1.
I have such log4j.xml

```
<?xml version="1.0" encoding="UTF-8"?>
<Configuration>
	<Appenders>
		<Console type="Console" name="ConsoleAppender">
			<PatternLayout pattern="%d{yyyy-MM-dd HH:mm:ss}  %-5p  %C{1}.%M:%L  >>> %m%n"/>
		</Console>
		<File name="FileAppender" fileName="${sys:myapp.saveDirectory}/${date:yyyy-MM-dd_hh-mm-ss}.log" append="false" immediateFlush="false">
			<PatternLayout>
				<header>------- ${sys:myapp.header} ------------------------------%n</header>
				<pattern>%d{yyyy-MM-dd HH:mm:ss}  %-5p  %C{1}.%M:%L  >>> %m%n</pattern>
			</PatternLayout>
		</File>
	</Appenders>
	
	<Loggers>
		<Logger name="console_only" level="warn">
			<AppenderRef ref="ConsoleAppender"/>
		</Logger>
		<Logger name="console_info" level="warn">
			<AppenderRef ref="ConsoleAppender"/>
		</Logger>
		<Logger name="file_error" level="error">
			<AppenderRef ref="FileAppender"/>
		</Logger>
		<Logger name="file_info" level="info">
			<AppenderRef ref="FileAppender"/>
		</Logger>
		<Root level="info">
			<AppenderRef ref="ConsoleAppender"/>
			<AppenderRef ref="FileAppender"/>
		</Root>
	</Loggers>
</Configuration>
```

and Log class is:
```

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

public class Log implements Thread.UncaughtExceptionHandler
{
        private static Logger log4j = LogManager.getLogger();
	static
	{
		System.setProperty("log4j2.*.Configuration.File", "/data/log4j2.xml");
		System.setProperty("myapp.saveDirectory", "some_dir");
		System.setProperty("myapp.header", "some_header");
		System.setProperty("log4j2.debug", "true");
	}

	public enum Appenders
	{
		// theese appenders should be present in 'log4j.xml'
		CONSOLE_ONLY("console_only"), CONSOLE_INFO("console_info"), FILE_INFO("file_info"), FILE_ERROR("file_error");

		private String name;

		private Appenders(String name)
		{
			this.name = name;
		}

		public String toString()
		{
			return name;
		}
	}

	public static Logger getLogger()
	{
		return log4j;
	}

	public static void setLogAppender(Appenders appender)
	{
		log4j = LogManager.getLogger(appender.toString());
	}

	@Override
	public void uncaughtException(Thread t, Throwable e)
	{
		e.printStackTrace();
		log4j.fatal("thread: " + t.getName(), e);
	}
}
```
For (very) earlier version this setup worked, but now I am even not sure if file log4j.xml is downoaded.
I don't see anything in console and no log files are created.
What I missed?

**Comment by ppkarwasz** [comment, marked as answer]
If you call the configuration file `log4j2.xml` and place it on the classpath, it will be automatically detected.

Otherwise you need to use a system property:

 - for 2.x the property is `log4j2.configurationFile`,
 - for 3.x the property is `log4j2.*.Configuration.file`,
 - [the one you set, `log4j.configuration`, was used in Log4j 1.x].

**Comment by riwarek** [comment]
Are all those properties documented somwhere?
And is 'log4j2.xml' name mandatory or it can be 'anyName.xml'?

**Reply by vy** [reply]
@riwarek, you can use `log4j2.debug` system property for troubleshooting Log4j issues. For instance, _"Is my Log4j configuration loaded?"_ etc. The properties @ppkarwasz shared are described in [the website](https://logging.apache.org/log4j/2.x/manual/configuration.html).

**Reply by riwarek** [reply]
I've found it here
https://logging.apache.org/log4j/2.x/manual/configuration.html#SystemProperties

**Reply by ppkarwasz** [reply]
The properties are documented on the respective web sites:

 - [Log4j 2.x properties](https://logging.apache.org/log4j/2.x/manual/configuration#SystemProperties),
 - [Log4j 3.x properties](https://logging.apache.org/log4j/3.x/manual/systemproperties).

The name of the file does not matter, only its extension needs to be compatible with the configuration format.

**Comment by riwarek** [comment]
What about passing params to xml?
I had:
`System.setProperty("log4j2.*.own.saveDirectory", Config.LOGS_DIR);`
I try something like
`<File  name="FileAppender" fileName="${sys:log4j2.own.saveDirectory}/${date:yyyy-MM-dd_hh-mm-ss}.log" append="false" immediateFlush="true">`

And I don't see any folder that I expect in the desired place.

**Reply by ppkarwasz** [reply]
The system properties lookup `${sys:<key>}` returns the value of the Java system property `<key>`. If you called the property `log4j2.*.own.saveDirectory`, you need to use:

```
${sys:log4j2.*.own.saveDirectory}
```

**Reply by riwarek** [reply]
I changed everything to form you've proposed (see first post) and I don't see anything in console (I try to call .
What I've observed is in this stacktrace:
```
Logger$PrivateConfig.filter(Level, Marker, String, Throwable) line: 523	
Logger.isEnabled(Level, Marker, String, Throwable) line: 173	
Logger(AbstractLogger).logIfEnabled(String, Level, Marker, String, Throwable) line: 2466	
Logger(AbstractLogger).warn(String) line: 3347	
SCoLog.handleInputParameters(String[]) line: 93	
SCoLog.main(String[]) line: 33	
```
```
  boolean filter(final Level level, final Marker marker, final String msg, final Throwable t) {
            final Filter filter = config.getFilter();
            if (filter != null) {
                final Filter.Result r = filter.filter(logger, level, marker, (Object) msg, t);
                if (r != Filter.Result.NEUTRAL) {
                    return r == Filter.Result.ACCEPT;
                }
            }
            return level != null && intLevel >= level.intLevel();
        }
```
`filter` is null

which caused this if `isEnabled` is `false`:
```
 public void logIfEnabled(
            final String fqcn, final Level level, final Marker marker, final String message, final Throwable t) {
        if (isEnabled(level, marker, message, t)) {
            logMessage(fqcn, level, marker, message, t);
        }
    }
 ```
Params of `isEnabled` are `WARN, null, (I_MAIN_PROG_5) Input parameters: cilog filog, null)`.

**Reply by ppkarwasz** [reply]
You have a couple of typos in your code. You set the property `log4j2.*.own.saveDirectory` in your code, but use `${sys:log4j.*.own.saveDirectory}` in the configuration file (without `2`).

BTW: you can name your system properties whatever you want, e.g. `own.saveDirectory`. The `log4j2.*` prefix is only required by the system properties recognized by Log4j 3.x.

**Reply by riwarek** [reply]
You are righ. This is the typo only on github. In app I had no mistakes. 
Anyway I simplified naming base your hint.
It doesn't impact on final result.

But now I think maybe I call loggers in wrong way?
I call it like this:
`Log.getLogger().info(LogString.I_MAIN_PROG_5 + arg);`
Then which logger will be taken? Root one or maybe `file_info` or `console_info`?

**Reply by riwarek** [reply]
Obviously there was a problem with finding log4j2.xml file.
This code helps:
```
LoggerContext context = (org.apache.logging.log4j.core.LoggerContext) LogManager.getContext(false);
File file = new File("path/to/a/different/log4j2.xml");

// this will force a reconfiguration
context.setConfigLocation(file.toURI());
```
I've found it here
https://logging.apache.org/log4j/3.x/faq.html

Now I see output in console.

**Reply by ppkarwasz** [reply]
I would advise against such a programmatic solution because:

1. It is specific to Log4j Core (unchecked cast). If a user replaces `log4j-core` with `log4j-to-slf4j` everything starts blowing up,
2. Setting `-Dlog4j2.*.Configuration.file=path/to/a/different/log4j2.xml` or `-Dlog4j2.configurationFile=path/to/a/different/log4j2.xml` works as well.

You can add `-Dlog4j2.debug` to your JVM options to debug any initialization errors.

**Comment by riwarek** [comment]
Additional proposition.
Currently, if I want to log something basic eg. application run parameter (only to console, only ot file, to file and to console). As I understand I have to create another config in xml.
It would be more convenient to add somethng like profile and pick different <Configuration> sets contained by one xml based on some parameter.

**Reply by ppkarwasz** [reply]
What you describe exists and is called [arbiter](https://logging.apache.org/log4j/2.x/manual/configuration.html#arbiters) in Log4j.

**Reply by riwarek** [reply]
Great!
I think that was the last thing.
Thank you for your patience and all explanations. I appreciate it.
